tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp

```diff
@@ -0,0 +1,546 @@
+      GeneralizedHarmonic::CharacteristicSpeedsCompute<Dim, Frame>::function(
+          gamma_1, lapse, shift, normal));
+}
+
+template <size_t Dim, typename Frame>
+void test_characteristic_speeds() noexcept {
+  const DataVector used_for_size(5);
+  pypp::check_with_random_values<1>(
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UPsi<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_upsi", {{{-10.0, 10.0}}}, used_for_size);
+  pypp::check_with_random_values<1>(
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UZero<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_uzero", {{{-10.0, 10.0}}}, used_for_size);
+  pypp::check_with_random_values<1>(
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UMinus<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_uminus", {{{-10.0, 10.0}}}, used_for_size);
+  pypp::check_with_random_values<1>(
+      compute_speed_with_tag<
+          Tags::CharSpeed<GeneralizedHarmonic::Tags::UPlus<Dim, Frame>>, Dim,
+          Frame>,
+      "TestFunctions", "char_speed_uplus", {{{-10.0, 10.0}}}, used_for_size);
+}
+
+// Test return-by-reference GH char speeds by comparing to Kerr-Schild
+template <typename Solution>
+void test_characteristic_speeds_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound) noexcept {
+  // Set up grid
+  const size_t spatial_dim = 3;
+  const size_t data_size = pow<spatial_dim>(grid_size_each_dimension);
+  Mesh<spatial_dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+                         Spectral::Quadrature::GaussLobatto};
+
+  using Affine = CoordinateMaps::Affine;
+  using Affine3D = CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = std::numeric_limits<double>::signaling_NaN();
+
+  // Evaluate analytic solution
+  const auto vars =
+      solution.variables(x, t, typename Solution::template tags<DataVector>{});
+  const auto& lapse = get<gr::Tags::Lapse<>>(vars);
+  const auto& shift = get<gr::Tags::Shift<spatial_dim>>(vars);
+  const auto& spatial_metric = get<gr::Tags::SpatialMetric<spatial_dim>>(vars);
+
+  // Get ingredients
+  const auto& gamma_1 = make_with_value<Scalar<DataVector>>(x, 0.1);
+  const auto& inverse_spatial_metric =
+      determinant_and_inverse(spatial_metric).second;
```

Comment:
I'd have thought that the lifetime of the rvalue temporary created for the returned value `determinant_and_inverse().second` should be extended by having bound it to a `const` lvalue reference.

Reply:
Just found a precedent for this: [here](https://github.com/sxs-collaboration/spectre/blob/develop/tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp#L116)

Reply:
I'm in favor of not making it a reference because:

1. The rules are tricky around lifetime of objects return from functions and bound to references.
2. It's misleading since a copy or move does actually need to occur. While the current code may look faster, it's not (but also not slower).

Reply:
The way this is now is good, but I decided to puzzle through the rules to figure out what is valid.

The original form `const auto& inverse_spatial_metric = determinant_and_inverse(spatial_metric).second;` was legal.  I was worried about the fact that this was binding to the inverse instead of the full det-inv pair, but it looks like that is fine.  (12.2.5 in @prayush's reference explicitly mentions subobjects.)

Binding a reference to the return value of a return-by-value function is quite common (although I personally find it questionable style-wise when not in generic code).  I believe the bullet @prayush refers to above is stating that code such as `const T& f() { return T{}; }` is bad.  The example at the bottom of the page shows the lifetime of a returned object being extended.